Enable bootscript templating

djurny opened the pull request:
- Enable bootscript templating using `envsubst` for bootscripts with the filename extension `.template`.
- Add generic bootscript. Plan is to have boards use this generic bootscript one by one.

# Description

In order to prepare for a generic bootscript, introduce simple templating for bootscripts using [`envsubst`](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html). The templating is needed as some things in the generic bootscript do require customization, where templating offers easy customization using already-existing shell variables. For example, `load_addr`, serial consoles, memory alignment settings and `family`/`vendor` names to enable proper name resolving for DT and DT overlay files.

# Documentation summary for feature / change

If bootscript name ends with the `.template` extension (could not find a more suitable extension), `envsubst` will be invoked to render the bootscript [template]. Rendering will consist of replacing any occurrance of a variable named `$BOOTSCRIPT_TEMPLATE__xxx` with the value in the environment of the build system at the time of rendering, in the bootscript template.

Some things to take into consideration, `BOOTSCRIPT` in the `boards/config/xxx` files can be overridden by `sources/families/xxx` or `sources/families/includes/xxx`. For example, for `boards/config/orangepizero.csc`, the setting for `BOOTSCRIPT` will be overridden by `sources/families/includes/sunxi-common.inc`. This can be solved with the following construct in `sources/families/includes/sunxi-common.inc`:
```
declare -g BOOTSCRIPT="${BOOTSCRIPT:-boot-sunxi.cmd}:boot.cmd"
```
For every board, this has to be inspected to make sure the bootscript selection is done as expected.

# How Has This Been Tested?

- [x ] Orangepi Zero
   - Change `BOOTSCRIPT` declaration in `sources/families/includes/sunxi-common.inc`.
   - Add the following to 'boards/config/orangepizero.csc`:
      ```
      DISPLAYCON=''
      SERIALCON="ttyS0:115200,ttyGS0"

      BOOTSCRIPT='boot-generic.cmd.template:boot.cmd'
      BOOTSCRIPT_TEMPLATE__ALIGN_TO='0x00001000'
      BOOTSCRIPT_TEMPLATE__BOARD_FAMILY="${BOARDFAMILY:-sun8i}"
      BOOTSCRIPT_TEMPLATE__BOARD_VENDOR='allwinner'
      BOOTSCRIPT_TEMPLATE__DISPLAY_CONSOLE='' # determine later on
      BOOTSCRIPT_TEMPLATE__LOAD_ADDR='0x45000000'
      BOOTSCRIPT_TEMPLATE__ROOTFS_TYPE="${ROOTFS_TYPE:-ext4}"
      BOOTSCRIPT_TEMPLATE__SERIAL_CONSOLE='' # determine later on
      ```
   - Compile for Orangepizero board and inspect generated (overridden and modified) `boot/boot.cmd`:
      ```
      [🐳|🌱] Rendering boot script template [ /armbian/config/bootscripts/boot-generic.cmd.template -> /armbian/.tmp/work-83d4164d-920e-4928-8d37-e7fa9aab818e/deb-bsp-cli-dUPmB/usr/share/armbian/boot.cmd ]
      [🐳|🔨]   '/armbian/config/bootenv/sunxi.txt' -> '/armbian/.tmp/work-83d4164d-920e-4928-8d37-e7fa9aab818e/deb-bsp-cli-dUPmB/usr/share/armbian/armbianEnv.txt'
      ...
      [🐳|🌱] Rendering boot script template [ /armbian/config/bootscripts/boot-generic.cmd.template ]
      [🐳|🔨]   '/armbian/config/bootenv/sunxi.txt' -> '/armbian/.tmp/rootfs-83d4164d-920e-4928-8d37-e7fa9aab818e/boot/armbianEnv.txt'
      ...
      ```
    - Resulting bootscript:
      ```
      # DO NOT EDIT THIS FILE
      #
      # Please edit /boot/armbianEnv.txt to set supported parameters
      #

      # This bootscript was generated on Sat, 21 Jun 2025 17:31:29 +0000

      # NOTE
      #       This file is the result of rendering a template using 'envsubst'.
      #       Any variable starting with 'BOOTSCRIPT_TEMPLATE__' is expected to be rendered.
      #       The build system will check if any variable is left after rendering and will
      #       exit with error if any unrendered variables are found.
      #       Check the following files for more details:
      #               lib/functions/bsp/armbian-bsp-cli-deb.sh
      #               lib/functions/rootfs/distro-agnostic.sh

      # NOTE
      #       If you intend to use 'outside' of 'global' variables from U-Boot, make sure that you do not change them!
      #       The boot logic will attempt a list of 'boot_targets' that all might rely on (environment) variables that
      #       have been set by U-Boot, either compile-time or as part of U-Boot's default 'bootcmd'.
      #       Any variable that this bootscript uses needs to be set explicitly and not conflict with any pre-set var>
      #       Variables that we might change will be saved in preset_x and variables we use will be copied into l_x.

      # default environment variables
      setenv align_overlap_oboe_avoidance "on"
      setenv align_to "0x00001000"
      setenv consoleargs_display ""
      setenv consoleargs_serial "console=ttyS0,115200 console=ttyGS0"
      setenv console "both"
      setenv docker_optimizations "on"
      setenv earlycon "off"
      setenv exit_on_critical_errors "on"
      setenv family "sun8i"
      setenv fdt_extrasize "0x00010000"
      setenv kver
      setenv load_addr_calc
      setenv overlay_error "false"
      setenv preset_fdtdir "${fdtdir}"
      setenv preset_fdtfile "${fdtfile}"
      setenv preset_kernel_comp_addr_r "${kernel_comp_addr_r}"
      setenv preset_kernel_comp_size "${kernel_comp_size}"
      setenv rootdev "/dev/mmcblk${devnum}p${distro_bootpart}"
      setenv rootfstype "ext4"
      setenv vendor "allwinner"
      setenv verbosity "1"

      # load addresses
      setenv load_addr "0x45000000"

      # NOTE
      #       No 'BOOTSCRIPT_TEMPLATE__' references beyond this point.
      ```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

djurny commented on lib/functions/rootfs/distro-agnostic.sh:
```diff
@@ -7,6 +7,70 @@
+	export BOOTSCRIPT_TEMPLATE__SERIAL_CONSOLE="${CONSOLEARGS:-}"
+}
+
+function render_bootscript_template() { (
+	typeset BOOTSCRIPT_TEMPLATE__CREATE_DATE
+	typeset SHELL_FORMAT
+
+	BOOTSCRIPT_TEMPLATE__CREATE_DATE="$(date -Ru)"
+
+	bootscript_export_display_console
+	bootscript_export_serial_console
+
+	SHELL_FORMAT="$(set | sed -En '/^BOOTSCRIPT_TEMPLATE__/ { s/=.*$//; s/^/$/; p; }')"
+	display_alert "Bootscript template variables to be rendered" "${SHELL_FORMAT:-N/A}" "debug"
+
+	export $(set | sed -En '/^BOOTSCRIPT_TEMPLATE__/s/=.*$//p')
+	envsubst "'${SHELL_FORMAT}'"
+) }
```
Hi bot,
The command `envsubst` is part of `gettext`.
To make sure `gettext` is is available, it was added to the list of packages to be installed in  `adaptative_prepare_host_dependencies` found in `lib/functions/host/prepare-host.sh`.
Groetjes,

igorpecovnik commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+	setenv bootargs "${bootargs} cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory"
+fi
+
+if test "${vendor}" == "allwinner" ; then
+	if test "${disp_mem_reserves}" = "off" ; then
+		setenv bootargs "${bootargs} sunxi_ve_mem_reserve=0 sunxi_g2d_mem_reserve=0 sunxi_fb_mem_reserve=16"
+	fi
+fi
+if test "${vendor}" == "marvell" ; then
+	#
+fi
+if test "${vendor}" == "rockchip" ; then
+	#
+fi
```
@djurny Is Rabbit correct about this?

djurny replied:
Hi @igorpecovnik,
Sorry, just saw your remark! I checked the U-Boot versions currently used and found that `do_test` will do a `strcmp` of the operator found against a struct containing supported operators. This `strcmp` seems to fail indeed when `==` is used instead of `=`. Also checked this on the U-Boot monitor, it deceptively returns `true` for anything you throw at it:
```
=> test "aaa" = "aaa" && echo yes
yes
=> test "aaa" == "aaa" && echo yes
yes
=> test "aaa" === "aaa" && echo yes
yes
=> test "aaa" ==! "aaa" && echo yes
yes
=> test "aaa" =! "aaa" && echo yes 
yes
=> test "aaa" =~ "aaa" && echo yes
yes
=> test "aaa" =~ "aaa" && echo yes
yes
=> test "aaa" =~ "aab" && echo yes
yes
=> test "aaa" = "aab" && echo yes 
=> test "aaa" == "aab" && echo yes
yes
=> test "aaa" == "" && echo yes   
yes
=> test "aaa" = "" && echo yes 
```
Apart from this indeed being incorrect, it's also not following the rest of the generic bootscript, where a single `=` is used.
In short, bot was right, @djurny made a booboo.
Will amend!

adeepn commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+	setenv bootargs "${bootargs} cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory"
+fi
+
+if test "${vendor}" == "allwinner" ; then
```
bloating... maybe better use any `include`  variant? envsubst from family config for example

djurny replied:
Hi @adeepn,
Thanks for your remark!
Actually thought of this as well, to make it a bit more flexible. A bit like how the DT fixup scripts are currently working. This will make it all a little more complicated though. Idea was to have something basic in first, make sure it works (for everyone that wants to use it), then improve bit by bit.
Groetjes,

djurny commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+		setexpr l_addr_next ${l_addr_next} + 1 ;
+		setexpr l_addr_next ${l_addr_next} * ${align_to} ;
+	fi'
+setenv func_inform 'test "${verbosity}" = "" || itest ${verbosity} -gt 0 && echo "${l_message}"'
```
Hi bot,
This is intended behavior. If verbosity is '0' or not set, it should stay quiet.
Doei.

djurny commented on lib/functions/rootfs/distro-agnostic.sh:
```diff
@@ -7,6 +7,70 @@
+	bootscript_export_display_console
+	bootscript_export_serial_console
+
+	SHELL_FORMAT="$(set | sed -En '/^BOOTSCRIPT_TEMPLATE__/ { s/=.*$//; s/^/$/; p; }')"
+	display_alert "Bootscript template variables to be rendered" "${SHELL_FORMAT:-N/A}" "debug"
+
+	export $(set | sed -En '/^BOOTSCRIPT_TEMPLATE__/s/=.*$//p')
+	envsubst "'${SHELL_FORMAT}'"
+) }
```
Hi bot,
The single quotes are not unnecessary. Please try this out yourself and see the results.
Doei.

djurny commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+fi
+
+# attempt to prepare for kernel address space randomization
+if kaslrseed ; then
+else
+	setenv l_message "Not able to prepare for KASLR."
+	run func_inform
+fi
```
Hi bot,
Nope. Please try this yourself. Also see for yourself if the `:` is supported on U-Boot's hush shell.
On a sidenote, if my memory serves me correctly, you advised me against adding an empty `;` ?
Doei.

djurny commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+# 	No 'BOOTSCRIPT_TEMPLATE__' references beyond this point.
+
+# environment run variables
+setenv func_align_addr_next '
+	test "${align_overlap_oboe_avoidance}" = "on" && setexpr l_addr_next ${l_addr_next} + 1 ;
+	setexpr l_modulo ${l_addr_next} % ${align_to} ;
+	if itest ${l_modulo} -gt 0 ; then
+		setexpr l_addr_next ${l_addr_next} / ${align_to} ;
+		setexpr l_addr_next ${l_addr_next} + 1 ;
+		setexpr l_addr_next ${l_addr_next} * ${align_to} ;
+	fi'
```
Hi bot,
Would love to do input sanitizing for all the things, but have to draw a line somewhere. The `align_to` can be overruled by `armbianEnv.txt`, but I must assume that if someone sets this, they should know what they are doing. If there is a fully interactive, always visible interface for the user where they can see this error pop up, it might be worthwhile, but also that is not the case.
Doei.

leggewie reviewed (commented):
Looks really awesome. I like the idea and hope it gets merged soon.

djurny commented:
Accidentally closed the PR due to my wonderfully oversensitive touchpad.

leggewie commented:
@djurny  I would say, this needs to be documented.  You already have a documentation section up there.  Can you prepare a PR for docs.armbian.com as well?

adeepn commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+
+# default environment variables
+setenv align_overlap_oboe_avoidance "on"
+setenv align_to "${BOOTSCRIPT_TEMPLATE__ALIGN_TO}"
```
BOOTSCRIPT_TEMPLATE__ALIGN_TO not defined anywhere

djurny replied:
Hi @adeepn,
That is correct. Due to the previous comments, I prepared the PR to just add the functionality, not the "roll-out" of it. 
If for example, the maintainer if Orangepi Zero wants to start using the template, the following would need to be added to `'boards/config/orangepizero.csc`:
```
    DISPLAYCON=''
    SERIALCON="ttyS0:115200,ttyGS0"

    BOOTSCRIPT='boot-generic.cmd.template:boot.cmd'
    BOOTSCRIPT_TEMPLATE__ALIGN_TO='0x00001000'
    BOOTSCRIPT_TEMPLATE__BOARD_FAMILY="${BOARDFAMILY:-sun8i}"
    BOOTSCRIPT_TEMPLATE__BOARD_VENDOR='allwinner'
    BOOTSCRIPT_TEMPLATE__DISPLAY_CONSOLE='' # determine later on
    BOOTSCRIPT_TEMPLATE__LOAD_ADDR='0x45000000'
    BOOTSCRIPT_TEMPLATE__ROOTFS_TYPE="${ROOTFS_TYPE:-ext4}"
    BOOTSCRIPT_TEMPLATE__SERIAL_CONSOLE='' # determine later on
```

See the PR description where it's laid out in a little more detail.

Hope that answers your question.
Groetjes,

adeepn replied:
it used in line 56:
`setexpr l_modulo ${l_addr_next} % ${align_to} ;`
so you need an some non-zero default there.

djurny replied:
Hi @adeepn,
It will be the responsibility of the board maintainer to make sure that the `BOOTSCRIPT_TEMPLATE__ALIGN_TO` variable is filled out to a value that makes sense. The build process will attempt to render the bootscript template and will also error out when unrendered variables are detected after rendering.
I do agree a sensible default value will be the better approach, but to be consistent, all variables should receive a sensible default. As hush and `gettext` are used to set/render/resolve all the `BOOTSCRIPT_TEMPLATE__***` variables, there is no easy construct to assign or resolve into a sensible default easily, e.g.
```
setexpr l_modulo ${l_addr_next} % ${align_to:-0x1000}
```
or
```
setenv align_to "${BOOTSCRIPT_TEMPLATE__ALIGN_TO:-0x1000}"
```
is not possible in the hush implementation in U-Boot (as far as I am aware). 
I'm all for making things unbreakable, but I'm also for the pragmatic approach. Not many people will (or should!) change these variables - whether it be in the board configuration or thru the `armbianEnv.txt` environment. Board maintainers are the intended userbase/audience for these variables.
Perhaps I'm missing some alternatives, so I'll have a look to see if something simple can be added at a later time.
Groetjes,

adeepn commented on config/bootscripts/boot-generic.cmd.template:
```diff
@@ -0,0 +1,537 @@
+	fi
+fi
+
+# booting failed, restore environment variables that are not unique
```
I don't think this is necessary: U-Boot does not save variables automatically without an explicit command, and for quick debugging it makes sense to keep the old values.
As for the original values—I think it's better to just use clearly different var names to avoid any confusion with the originals.

adeepn replied:
> it's about the `distro_bootcmd` loop,

Hmm, I assumed that everything would stop at this script and after that we'd just end up in a regular shell. If it actually tries the next option, then I definitely agree.


> some variables are used by U-Boot commands implicitly

bootz not using implicitly kernel_comp_addr_r/kernel_comp_size_r :
```
bootz [<addr> [<initrd>[:<size>]] [<fdt>]]
addr
  address of kernel image, defaults to the value of the environment variable $loadaddr.

initrd
  address of the initial RAM disk. Use ‘-’ to boot a kernel with a device tree but without an initial RAM disk.
```
https://docs.u-boot.org/en/stable/usage/cmd/bootz.html

djurny replied:
> > some variables are used by U-Boot commands implicitly
> bootz not using implicitly kernel_comp_addr_r/kernel_comp_size_r :
> ```
> bootz [<addr> [<initrd>[:<size>]] [<fdt>]]
> addr
>   address of kernel image, defaults to the value of the environment variable $loadaddr.
> https://docs.u-boot.org/en/stable/usage/cmd/bootz.html

Hi @adeepn,
You have a point: After re-reading the code, seems that U-Boot `booti` command implicitly uses the `kernel_comp_addr_r` and `kernel_comp_size_r`, instead of `bootz` as I mentioned.

`zImage` is a self-extracting version of the kernel, which needs `bootz` and `Image` needs `booti` which also can handle (non-self-extracting) compressed images. These compressed images need the two variables set to declare a memory area where the decompressor can use.

https://github.com/u-boot/u-boot/blob/master/doc/usage/cmd/booti.rst
https://github.com/u-boot/u-boot/blob/29d5636c6f473ac2cf273a2fd07981125197564f/cmd/booti.c#L50

Groetjes,

adeepn reviewed (commented):
At first, I was against the excessive complexity, but there are some interesting ideas here.
For example, calculating the lengths of the kernel, FDT, and ramdisk.
Overall, it actually turns out to be a fairly sound idea. 

But at first glance, what's missing here is:
- support for fit-image (bootm command)
- extending the func_inform function to debug/info levels in order to separate useful information from pure debugging (for example, printing bootargs line by line)
We can start with this and improve it further later.

leggewie reviewed (changes requested):
I am really looking forward to this change, thank you for your work, but please let's make sure this is documented before this gets merged

The-going commented:
warning:
We have a [configng](https://github.com/armbian/configng) utility that provides overlay management in particular.
Basically, this control consists of two parts of variable names: `${overlay_prefix}-${overlay_file}.dtbo`
But in the vendor kernel for rockchip-rk3588 there are overlays whose names contain only `${overlay_file}.dtbo`

In order for the utility to work correctly, we had to add a lexical analyzer [configng/tools/modules/system/manage_dtoverlays.sh](https://github.com/armbian/configng/blob/25d6fba8b6ced9eba00d2dc2e76c0b58634475f7/tools/modules/system/manage_dtoverlays.sh#L36-L100)
And we understand that variable names matter a lot here.

igorpecovnik commented:
Is it possible to complete this within next two weeks?

djurny commented:
Hi all,
Still adding more text, but would like to get some human feedback if this is going in the right direction. 
Thx,
Groetjes,

djurny commented:
Added a story on how to use the generic bootscript template in `boot-generic.cmd.template.md`.
Added a dedup action on the `DISPLAYCON` and `SERIALCON` variables.

igorpecovnik commented:
Something weird happened here.

djurny commented:
> Something weird happened here.

I think this is about me and my habits of using continuous/cmsynergy/clearcase/subversion. Somehow I keep on getting into situations with git that make my head spin.

I'm going to close this PR and make a fresh one.

igorpecovnik commented:
No worries, ok.
